### DISCUSSION TITLE
Librocco app/data flows suggestions/discussion

### DISCUSSION THREAD

**Discussion by ikusteu** [librocco/librocco#85] [General]
I this discussion, I would like to tackle the app logic and data flows for the Librocco app. A lot of the stuff included here (split by sections) should probably be their own separate units of work (and we should definitely split those later), but I've included all of them in this discussion as a lot of the data requirements depend on defining some of this logic.

# TOC

1. [Terminology](#1-terminology)
2. [Views](#2-views)
3. [Search](#3-search)
4. [Frontend data flows](#4-frontend-data-flows)

# 1) Terminology

Before getting into the matter, I would like to define terminology used in this "spec" as a lot of the naming here can be ambiguous, or even confusing (view as in dashboard view vs view as couch db index). All of the nomenclature here is working terminology and scopes to this document with sole purpose of providing clarity for understanding/discussion.

## Visual elements:

- **SearchBar:**

  - well...self explanatory, but: a search bar at the top of the dashboard

- **SideNav:**

  - a navigation bar on the left hand side of the app, used to navigate between top level views
  - rendered as part of the app layout (wrapping the dashboard)

- **SubNav:**

  - a left hand side navigation **inside** the dashboard, rendered differently per each view (warehouses, notes or notes grouped by warehouses)
  - used to navigate between subviews or as buttons triggering an action

- **Table section:**

  - the right hand (large) section showing table of data entries: books in stock, books in current note, etc.

## Abstract:

- **view:**

  - when I say view, I mean different top level view of the client dashboard
  - had this not been a dashboard, "view" would be equivalent to a "page"
  - the "views" each represesnt a different route and are navigated between using the **SideNav**
  - **not to be confused with CouchDB views, when I refer to CouchDB views, I will state so explicitly**

- **subview:**

  - sub view rendered under a particular view: displaying single note in `in-notes` view, or stock in a particular warehouse (or "All" warehouses)

- **book:**

  - when I say a book, I mean a single unit, an instance of a book
  - e.g. Basic Economics, Samuelson & Nordhaus, in croatian, located on my bookshelf, and not any other instance of the same publication

- **publication:**
  - @silviot had a different term for this, but I can't remember it
  - a publication of a book, with its own ISBN, author, title, year, etc.
  - e.g. Basic Economics, Samuelson & Nordhaus, 18th edition, published by McGraw Hill, 2005.
  - not a book in a physical sense
  - there are numerous physical copies (books) of a given publication

# 2) Views

Here we have a general overview of all views and subviews of the dashboard. These are used as a starting point for deriving/defining the required functionality and data flows:

## Stock view (root)

- route: `/` (root, default)
- shows all books in stock: either in a particular warehouse, or in entire stock
- the entries (rows) are shown as **publications** with a property displaying quantity (number of copies in stock)
- allows searching (filtering using js-search) for each particular subview

- **subviews (control which stock should be shown):**

  - **"All"** (default):

    - table section: shows all book entries in stock
    - when searching: filters through all entries in stock (and displays filtered results in the table section, as **publications** with quantity)

  - **\<warehouse-name\>:**

    - table section: shows all books, in stock, for a particular warehouse
    - when searching: filters through all entries in that warehouse's stock (and displays filtered results in the table section, as **publications** with quantity)

## In-note view

- route: `/in-notes` (I don't really like this, suggestions are more than welcome)
- used to inspect all the past **in-notes**, add new and commit existing (draft) **in-notes**
- allows searching through all of the **in-notes**

- **subviews (control which note should be shown):**

  - **default:**
    - no note is being inspected, nothing is rendered in the table section
    - all notes are displayed in the **SubNav**, groupped by warehouse
  - **\<note-name\>:**
    - in table section: all books in a particular note: as **books**, physical copies, not publications (two copies of the same book are shown as two entries)
    - when searching: opens search view
  - **new-note/draft:**
    - initially it opens a drawer to add books to the note, but when the drawer is closed, the note persists
    - should look the same as a regular note subview :point_up: but should have additional actions (open drawer to add books, remove books, delete note)
    - navigating away from the new note should store is as draft for later edit/commit
    - additional actions should be removed when the note gets committed (the note becomes a regular committed note like the one above :point_up:)
  - **search** (a special view):

    - in table section:
      - unfiltered results _(no search string, wouldn't happen in reality as search view is triggered by search string being defined)_: displays all **books** from all **in-notes**
      - search results: all **books** from all **in-notes**, filtered by `js-search` with respect to the search string
    - when searching:
      - filters the results in the table section
      - filters "empty" notes in **SubNav**:
        - the notes which don't contain any of the books shown in filtered results are considerd "empty" in this case
        - thus, only the notes containing at least one of the books in filtered results are shown in **SubNav**
      - clicking on a book in table section opens the full **in-note** that the book belongs to (including books not matched by the filter, but included in the note)
      - clicking on a note in **SubNav** opens the full **in-note** in table section (including books not matched by the filter, but belonging to the note)
      - search state (and filtered results) persist until the search string is changed or cleared
      - the last should be assessed in terms of display

## Out-note view

- route: `/out-notes`
- see [in-note view](#in-note-view)
- only difference from the **in-note** view is that the notes in **SubNav** aren't grouped by the warehouse

# 3) Search

There's been talk about search and how search should behave in different views. My suggestion is that we enable search in each view, but with slightly different logic (as I've briefly explained above):

- in all views full-text-search is used

## Search logic per view:

- **warehouse** view:

  - search is straightforward, the filter is applied to the entries (publications) already shown in the table section (with respect to the subview, see [views](#stock-view-root))

- **in-note** view:

  - search would create a special case (search) view in the table:

    - would search through all **books** in all **in-notes** and the results would be shown in the table section
    - the results displayed will not be books as mere publications, but rather as a single copy from a single note:

      - example:

        We have three **in-notes**:

        - **note-a:**
          - The Great Gatsby, F. Scott Fitzgerald
          - The Island of Dr. Moreau, H. G. Wells
        - **note-b:**
          - The Great Gatsby, F. Scott Fitzgerald
          - The Lord of the Rings: The Fellowship of the Ring, J.R.R. Tolkien
        - **note-c:**
          - Inferno, Dante
          - The Silmarillion, J.R.R. Tolkien

        Searching for `"Gatsby"` would show two results:

        - The Great Gatsby, F. Scott Fitzgerald (note-a)
        - The Great Gatsby, F. Scott Fitzgerald (note-b)

    - as books get filtered out, only the notes containing at least one of the books in search results would be shown in the **SubNav**, thus notes shown will get filtered out as the list narrows down, following the previous example:

      To start with, three notes are shown in the sub-nav:

      - **note-a**
      - **note-b**
      - **note-c**

      After searching for `"Tolkien"`, only the two notes will be shown in sub-nav:

      - **note-b** (contains "The Fellowship of the Ring", matched by `"Tolkien"`)
      - **note-c** (contains "The Silmarillion", matched by `"Tolkien"`)

      _**note-a** is filtered out as it doesn't contain any publication matched by `"Tolkien"`_

- **out-note** view:
  - same as **in-note** search

## Implementation and data passing

My suggestion was to make the search string available anywhere in the app by using the route query param `?s=<search-string>`, but considering Svelte and svelte's state management (and ease of use of stores and passing them around) I feel it would make the most sense to use the search store(s) for this:

- **search store:**
  - writable store (available to any component/store) used by **SearchBar** to update the search string
  - should exist for the lifetime of the app (indefinitely)
- **filter stores:**
  - derived stores in charge of filtering
  - each view would have its own filter store (as filtering logic differs per view)
  - the **filter store** would subscribe to **search store** (to get search string) and some other data source (as different views need different data)
  - lifetime of a **filter store**:
    - when a view is mounted (we've navigated to the view), a new store would be created (with logic specific to that view's search requirements)
    - when a view is destroyed (we've navigated away from the view), the store should also be destroyed to prevent the store from getting updated (in the background) on each search string change, even though we're not consuming the data from that store
  - with respect to the store's lifetime requirements, we should utilise the factory functions creating the stores, rather than importing the stores themselves

# 4) Frontend data flows

With respect to the views and search logic described above, I've derived data requirements (what should be provided by the backend) and actions/updates (how the FE can interact with the BE to update the data in the DB)

## Data requirements:

This is the data required by our views and our backend (db) should be structured to cater to those needs.

_At the time of writing of this section I haven't looked into CouchDB yet, so the following data is structured exactly as needed by FE and the implementation is TBD in the following sections._

### "Stock" view

- required data:
  - full stock data: for "All" subview
  - stock data per warehouse basis
- data fetched for each subview should include an array (or a record) of publication entries, each entry should look something like this:

  - **isbn**
  - **title**
  - **author**
  - **editor** - ?I'm unclear about this...I would expect publisher here?
  - **price** - ?unclear on how the price should be set?
  - **quantity** (available in stock) - @silviot mentioned this, it should be updated in the mockup somehow

### "In-notes" view

- required data:
  - per note basis - an **in-note** entry
  - for search purposes - all **books** contained in all **in-notes**
- both data sets would include an array (or a record) of book entries, book entries looking a lot like the publication entries:

  - **isbn**
  - **title**
  - **author**
  - **editor** - ?I'm unclear about this...I would expect publisher here?
  - **price** - ?unclear on how the price should be set?
  - **note:**

    - this property would allow for navigation to the note when the entry is clicked
    - additionally it could prove useful for filtering notes to display in the **SubNav** with respect to search results (as described in [search section](#search-logic-per-view)), something like:

      ```typescript
      const notesToShow = [
        ...searchResult.reduce((acc, { note }) => acc.add(note), new Set()),
      ].sort((a, b) => (a < b ? -1 : 1));
      ```

    - ideally, the logic mentioned above could be included with the search results to avoid excess iterations, but this works

  - **date** - the date the book has been taken in (essentially the date of the note itself, but it makes sense to include it in the book structure, for search result display puropses)

### "Out-notes" view:

- essentialy the same as in-notes, only with out-note entries

## Actions/updates

This section includes the actions dispatched from the client app to update the DB. All updates are handled through notes:

- **in-note** - accounts for in-flow of the books to the stock (the quatity in DB is incremented for each book)
- **out-note** - accounts for out-flow of the books from the stock (the quentity in DB is decremented for each book)

All notes and note-related actions are the same regardless of in/out note:

- **lifetime** of a note:
  - **new note** is created (by clicking a new note button in the dashboard):
    - the blank note is stored in the db in "draft" state
    - no **date** (time) is added to the note
    - no subsequent updates to the stock take place yet
  - **edit note** - books are added/removed to and from the note:
    - each change is updated to the db
    - the notes are editable as long as they're in the "draft" state
  - **commit note**
    - the note is committed and a transaction takes place
    - the stock is updated with respect to the note (this happens in the backend)
    - note get's locked for editing/removal as it has been committed
  - **remove note**
    - the note is removed (this can only be done if the note was not yet committed)

```mermaid
graph TD
    A(New note) -->|note is created in ''draft'' state| B(Add/remove books)
    B -->|delete note| C(Note is deleted)
    B -->|commit note| D(Note is persisted, no edits/deletes are available)
```

### Actions

```typescript
/**
 * Creates a new note ("in" or "out"):
 * - opens the note view and an shows an edit drawer (immediately)
 * - dispatches an update against the DB to register the new note (returning a promise)
 * @TODO we should make the note id accessible to the app (either created immediately, or returned from DB when the note is created)
 */
interface NewNote {
  (type: "in" | "out"): Promise<void>;
}
/**
 * Dispatches an update to the DB for the note: adding or removing books in the process.
 * @param noteId id of a particular note
 * @books all book entries in the updated note: "date" and "note" properties are omitted as they're assigned/aggregated by the server
 * when the note gets commited
 */
interface UpdateNote {
  (
    noteId: string,
    books: Omit<BookInterface, "date" | "note">[]
  ): Promise<void>;
}
/**
 * Dispatches a commit for the note, triggering the stock update (backend side) and persisting the note (no edit/remove is possible).
 * @param noteId
 */
interface CommitNote {
  (noteId: string): Promise<void>;
}
/**
 * Removes the (not yet committed) note from the DB.
 * @param noteId
 */
interface DeleteNote {
  (noteId: string): Promise<void>;
}
```

**Comment by ikusteu** [comment]
What we've discussed at the meeting:
- notes should be as similar to the warehouse as possible - each warehouse entry will show book data + quantity and the note entries should be structured the same (with additional `note` field pointing to the note they belong to)

**Comment by ikusteu** [comment]
**Update:** the update functions shouldn't concern themselves with opening a note in view (that should be handled by the UI logic).
With respect to that, I think it would make sense for `newNote` function to return the generated `id` of the note (for future use).

**Reply by fadwamahmoud** [reply]
Why not just keep uncommitted notes in local state instead of having two possibly unnecessary actions (creating empty note and removing it) 
Another thing, what if we [use current date](https://pouchdb.com/guides/bulk-operations.html#use-alldocs-to-read-many-docs) instead of a random id as the `noteId`, that would make it sorted by default and we wouldn't need a different `date` prop 
@ikusteu

**Reply by silviot** [reply]
> Why not just keep uncommitted notes in local state instead of having two possibly unnecessary actions (creating empty note and removing it)

@fadwamahmoud Fair objection! It's a requirement that draft notes need to be shared between devices: one should be able to create a note on device A, add some books, and then switch to device B to continue working on them.

> Another thing, what if we [use current date](https://pouchdb.com/guides/bulk-operations.html#use-alldocs-to-read-many-docs) instead of a random id as the noteId, that would make it sorted by default and we wouldn't need a different date prop

I haven't yet gone into the details of this, but I believe [a UUIDv1](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_1_(date-time_and_MAC_address)) would be a good ID and also provide info about note creation. It would also lexicographically sort by date.

One thing I'm not super-happy about UUIDs is their length. A quick search just unearthed the [`short-uuid`](https://github.com/oculus42/short-uuid) library. We might consider using it.

**Comment by silviot** [comment]
@ikusteu I believe `updateNote` should not take the whole state of the note to do the update.

When the user adds a book to a note (or removes one) they don't really mean "Make the state of this note be this list of 79 books with their quantities". They most likely mean "Increase the quantity of book `978XXXXXXXXXX` by 2 in this note. Or "Bring the quantity of book `978XXXXXXXXXX` down to 0 in this note. The UI element that will dispatch the action (in most cases a field where the user entered an ISBN code) is not aware of the current state of the note, unless we do some extra work.
Also, to minimize conflicts, we should strive to do writes that are as small as possible.

I propose a different API:

```
interface UpdateNote {
  (
    noteId: string,
    isbn: string,
    quantity: number
  ): Promise<void>;
}
```

**Reply by ikusteu** [reply]
Totally agree, I would even do something like `addBooks`, `removeBooks`, but let's, maybe, continue this in a newly created discussion about note acton definitions and db interface. #91
